_pages/automatic_differentiation.md

```diff
@@ -0,0 +1,153 @@
+---
+title: "Compiler Research Research Areas"
+layout: gridlay
+excerpt: "Automatic differentiation (AD) is a powerful technique for evaluating the
+derivatives of mathematical functions in C++, offering significant advantages
+over traditional differentiation methods."
+sitemap: true
+permalink: /automatic_differentiation
+---
+
+## Automatic differentiation
+
+Automatic differentiation (AD) is a technique for evaluating the derivatives
+of mathematical functions in C++, offering a number of advantages over
+traditional differentiation methods. By leveraging the principles of Automatic
+Differentiation, programmers can efficiently calculate partial derivatives of
+functions, opening up a range of applications in scientific computing and
+machine learning.
+
+### Understanding Differentiation in Computing
+
+Differentiation in calculus is the process of finding the rate of change of
+one quantity with respect to another. There are several principles and
+formulas for differentiation, such as Sum Rule, Product Rule, Quotient Rule,
+Constant Rule, and Chain Rule. For Automatic Differentiation, the Chain Rule
+of differential calculus is of special interest.
+
+Within the context of computing, there are various methods for
+differentiation:
+
+- **Manual Differentiation**: This consists of manually applying the rules of
+  differentiation to a given function. While straightforward, it can be
+  tedious and error-prone, especially for complex functions.
+
+- **Numerical Differentiation**: This method approximates the derivatives
+  using finite differences. It is relatively simple to implement, but can
+  suffer from numerical instability and inaccuracy in its results.
+
+- **Symbolic Differentiation**: This approach uses symbolic manipulation to
+compute derivatives analytically. It provides accurate results but can lead to
+lengthy expressions for large computations. It is limited to closed-form
+expressions; that is, it cannot process the control flow.
+
+- **Automatic Differentiation (AD)**: Automatic Differentiation is a highly
+  efficient technique that computes derivatives of mathematical functions by
+  applying differentiation rules to every arithmetic operation in the code.
+  Automatic Differentiation can be used in two modes: 
+
+    - Forward Mode: calculates derivatives with respect to a single variable, and 
+    
+    - Reverse Mode: calculates gradients with respect to all inputs
+    simultaneously.
+
+### Automatic Differentiation in C++
+
+Automated Differentiation implementations are based on Operator Overloading or
```

Comment:
this part needs work..both to add some links to more authoritative web pages and to focus on the source transformation method rather than just the operator overloading one.

Reply:
New description:
Compiler Research Group's focus has been on exploring the [Source Code Transformation] technique, which involves  constructing the computation graph and producing a derivative at compile time. 

[The source code transformation approach] enables optimization by retaining all the complex knowledge of the original source code. The compute graph is constructed before compilation and then transformed and compiled. It typically
uses a custom parser to build code representation and produce the transformed code. It is difficult to implement (especially in C++), but it is very efficient, since many computations and optimizations are done ahead of time.

Links:

[Source Code Transformation]: https://compiler-research.org/assets/presentations/V_Vassilev-SNL_Accelerating_Large_Workflows_Clad.pdf

[two major techniques]: https://compiler-research.org/assets/presentations/G_Singh-MODE3_Fast_Likelyhood_Calculations_RooFit.pdf

[The source code transformation approach]: https://compiler-research.org/assets/presentations/I_Ifrim-EuroAD21_GPU_AD.pdf